### ISSUE TITLE
Proposal: Smart Tabs rule for `indent`.

### ISSUE THREAD

**Issue by AcceptableIce** [eslint/eslint#9089]
**Tell us about your environment**

* **ESLint Version:** 4.2.0
* **Node Version:** 7.8.0
* **npm Version:** 5.0.3

**The problem you want to solve.**
The `smart-tabs` setting for the `no-mixed-spaces-and-tabs` rule causes a lot of "false positives" for us, as it doesn't ensure that the proper number of tabs occur before any spaces. For example"

```js
let test = function(arg) {
  if(arg) {
    return true;
  }
};
```

The `if` line has a tab, while the `return` line has a tab, followed by two spaces. I'd like it so that the `return` lab must have two tabs, then can have any number of alignment spaces.

This comes up in a lot of code reviews for me, where eslint says heavily indented code is correct, but it's actually a single tab followed by an insane number of spaces.

**Your take on the correct solution to problem.**
Add a `smart-tabs` or `strict-smart-tabs` option to `indent`, which requires the expected number of tabs before any indentation spaces are allowed.

**Comment by not-an-aardvark** [comment]
Thanks for the proposal.

> which requires the expected number of tabs before any indentation spaces are allowed.

Could you clarify what the "expected number of tabs" is in general? Is that what `indent` rule currently does with the `tab` setting (when there are no alignment spaces)?

Currently, the `indent` rule doesn't make a distinction between "indentation" whitespace and "alignment" whitespace -- any whitespace at the beginning of a line is considered "indentation".

**Comment by AcceptableIce** [comment]
Here's another example

```js
function test(var1, var2) {
	if(var1 === var2) {
	    return var1;
	} else {
		return var2;
	}
}
```

Assume we have `indent` set to expect tabs.

Notice how lines 3 and 5 have different indentation; we expect two "units" of indentation (two tabs, 4/8 spacesm, etc) on those lines. Line 3 has one tab followed by four spaces, while line 5 has two tabs. eslint treats line 3 as correct, even though it doesn't have the required two tabs of indentation.  My proposal is a rule that causes line 3 to be treated as incorrect, but if it had two tabs followed by some spaces, that would be acceptable under `smart-tabs`.

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get consensus from the team, so I'm closing it. We define [consensus](http://eslint.org/docs/maintainer-guide/issues.html#consensus) as having three :+1:s from team members, as well as a team member willing to champion the proposal. This is a high bar by design -- we can't realistically accept and maintain every feature request in the long term, so we only accept feature requests which are useful enough that there is consensus among the team that they're worth adding.

Since ESLint is pluggable and can load custom rules at runtime, the lack of consensus among the ESLint team doesn't need to be a blocker for you using this in your project, if you'd find it useful. It just means that you would need to [implement the rule](http://eslint.org/docs/developer-guide/working-with-rules) yourself, rather than using a bundled rule that is packaged with ESLint.
